src/audio.c

```diff
@@ -121,186 +122,290 @@ void record(void) {
+	return -1;
+    }
+
+    g_free(expanded_dir);
+
+    for (int i = 4; i < 15; i++)
+	clear_line(i);
+
+
+    if (n > 48) n = 48;	    /* limit number of file to display */
+
+    int i = 10;
+    int j = 4;
+    for (int k = 0; k < n; k ++) {
+
+	char *printname = strip_suffix(nameList[k] -> d_name);
+	mvprintw(j, i, "%s", printname);
+	g_free(printname);
+
+	i += 10;
+    	if (i > 60) {
+	    i = 10;
+	    j++;
+	}
+    }
+    refreshp();
+
+    return 0;
+}
+
+
+/* kill process (first command token in string) with SIGTERM */
+static void stop_command(char *string) {
+    gchar **vector = g_strsplit_set(string, " \t", 2);
+    char *command = g_strconcat("pkill -SIGTERM -n ", vector[0], NULL);
+    g_strfreev(vector);
+
+    IGNORE(system(command));
+    g_free(command);
+}
+
+
+/* voice recorder handling - recording and play back */
+static void vr_start(void) {
+    IGNORE(system("echo " " > ~/.VRlock"));
+
+    char *command = g_strconcat("mkdir -p ", soundlog_dir, "; ",
+	    "cd ", soundlog_dir, "; ",
+	    soundlog_record_cmd, " >/dev/null 2>/dev/null &", NULL);
+
+    IGNORE(system(command));
+    g_free(command);
+}
+
+
+static void vr_stop() {
+    IGNORE(system("rm ~/.VRlock"));
+    stop_command(soundlog_record_cmd);
+}
+
+
+/* substitute filename with appended file type suffix in play command and
+ * prepare for running the command in the soundlog directory
+ */
+static char *prepare_playback_command(char *filename) {
+    char *file = g_strconcat(filename, ".au", NULL);
+
+    GRegex *regex = g_regex_new("\\$1", 0, 0 , NULL);
+    char *play_command = g_regex_replace(regex, soundlog_play_cmd, -1, 0,
+	    file, 0, NULL);
+    g_regex_unref(regex);
+    g_free(file);
+
+    char *full_command = g_strconcat("cd ", soundlog_dir, "; ",
+		play_command, NULL);
+    g_free (play_command);
+
+    return full_command;
+}
+
+/* voice keyer handling - recording and playback */
+/*--------------------------------------------------------------------------*/
+static void vk_do_record(int message_nr) {
+
+    extern char ph_message[14][80];
+
+    mvprintw(15, 20, "recording %s", ph_message[message_nr]);
+    mvprintw(16, 20, "ESC to exit");
+    mvprintw(17, 20, "");
+    refreshp();
+
+    GRegex *regex = g_regex_new("\\$1", 0, 0 , NULL);
+    char *command = g_regex_replace(regex, vk_record_cmd, -1, 0,
+	    ph_message[message_nr], 0, NULL);
+    g_regex_unref(regex);
+
+    /* let the command run in background so we can stop recording by
+     * <esc> key later */
+    char *reccommand = g_strconcat( command, " &", NULL);
+
+    IGNORE(system(reccommand));
+    g_free(command);
+    g_free(reccommand);
+
+    /* Loop until <esc> key pressed */
+    while (1) {
+	if (key_get() == ESCAPE) {
+	    stop_command(vk_record_cmd);
+	    break;
 	}
     }
 }
 
 
-/* playing recorded voice keyer messages */
+/* playing recorded voice keyer messages in background */
 void *play_thread(void *ptr) {
     char *audiofile = (char *)ptr;
```

Comment:
`ptr` may not point to a local variable as it can be already unavailable when the thread actually starts running.
Currently this typically doesn't happen (all playable strings are globals) but for #316 this could change.
An option is to strdup the string in `vk_play_file` and the free it here after use.

Reply:
Ok. I will look into it.

Reply:
See commit https://github.com/Tlf/tlf/pull/318/commits/5a76f5cb833247137dfd42a1b952fa8d4c7cca33